### PR TITLE
ci: create test quality reports

### LINKED ISSUE #20: Report code coverage
* on PR via _Azure Pipelines_
* also publish test tesults

### LINKED ISSUE #21: Check via Mutation Testing
* impacted code on _PR_
* all code on push to _main_

### PR DESCRIPTION
**Pull request by Flash0ver**
Closes #20
Closes #21

_xUnit.net_ `Theory` published as just one single test - see microsoft/azure-pipelines-tasks#13842

### CODE REVIEW

**Review by thomas-pemsel — APPROVED** [review]

File: azure/pipelines/templates/code-coverage.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    inputs:
+      packageType: 'sdk'
+      useGlobalJson: true
+      performMultiLevelLookup: true
+  - task: DotNetCoreCLI@2
+    displayName: Install .NET Tools
+    inputs:
+      command: 'custom'
+      custom: 'tool'
+      arguments: 'restore --configfile $(NuGet_Configuration_File) --tool-manifest $(Tool_Manifest_File) --no-cache'
```
Comment (dev-in-disguise) [inline azure/pipelines/templates/code-coverage.yml:36]: I'm not sure if this bothers me but by having a single tool-manifest we restore all dotnet tools we use. Right now this leads to `stryker` being restored although it's not used in the code-coverage job. A shame that there seems to be no way of handling this better

**Review by dev-in-disguise — APPROVED** [review]

File: azure/pipelines/templates/code-coverage.yml (hunk 1 above)
Comment (thomas-pemsel) [reply azure/pipelines/templates/code-coverage.yml:36]: Just as a note for later discussions:
What about having multiple dotnet-tools.json with the "isRoot" set to false?
Can this be a way to achieve this?

**Review by thomas-pemsel — APPROVED** [review]

File: scripts/Test-Solution.ps1 [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+if (Test-Path -Path $TestResultsDirectory) {
+    Remove-Item -Path $TestResultsDirectory -Recurse
+}
+
+dotnet test $SolutionFile --configuration $Configuration --no-build --nologo --results-directory $TestResultsDirectory /p:CollectCoverage=true /p:CoverletOutputFormat=cobertura /p:CoverletOutput=./TestResults/
+
+dotnet tool run reportgenerator "-reports:$RepositoryRootPath/**/coverage.cobertura.xml" "-targetdir:$CoverageReportTargetDirectory" -reporttypes:HtmlInline_AzurePipelines
+
+$TestProjectFile = $($TestProjectFile.Replace('\', '/'))
+dotnet tool run dotnet-stryker --solution-path $SolutionFile --test-projects "['$TestProjectFile']" --config-file-path $StrykerConfigurationFile
```
Comment (thomas-pemsel) [inline scripts/Test-Solution.ps1:43]: I just tried to run this script in my other branch.
As a FYI: I downloaded the following files:
* scripts/Test-Solution.ps1
* .config/dotnet-tools.json
* nuget.config

First of all, I had to update Powershell/install a new Powershell version, as per default I only had Powershell 5.1.
The script required at least Powershell 6.0, and with the "official" Script `iex "& { $(irm https://aka.ms/install-powershell.ps1) } -UseMSI"`
This installed Powershell 7 💪 

And then, Stryker didn't run for me with many errors like this:
`[22:33:48 WRN] Stryker.NET encountered an compile error in C:\Users\[...]\F0.Minesweeper.Logic\Cell.cs (at 6:34) with message: The feature 'init-only setters' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version. (Source code: init;)`

As a solution, I added `--language-version preview` to the end of this command to specify a newer C# version.
Do we want to generally add this option/switch to this Stryker call? Let's discuss next time 😄

File: azure/pipelines/templates/code-coverage.yml (hunk 1 above)
Comment (Flash0ver) [reply azure/pipelines/templates/code-coverage.yml:36]: Separate `tool-manifest` files seems non-tool-manifesty.
Lest's completely parameterize all templates, including the respective .NET tool version (latest if not specified), and move those job-tempates to a separate repo, so everyone could resuse them.

File: scripts/Test-Solution.ps1 (hunk 2 above)
Comment (Flash0ver) [reply scripts/Test-Solution.ps1:43]: Cool, adding the `--language-version preview`

**Pull request merged**